fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/rules/DecorrelateValuesRule.java

```diff
@@ -0,0 +1,375 @@
+ *   ) f
+ *   WHERE f.c IS NULL
+ * }</pre>
+ *
+ * <p>
+ * The new child is also eligible for value decorrelation, leading to the expression:
+ * </p>
+ *
+ * <pre>{@code
+ * SELECT f.d
+ *   FROM (SELECT c, d
+ *           FROM T
+ *           WHERE a = 42 AND b = 'hello'
+ *   ) f
+ *   WHERE f.c IS NULL
+ * }</pre>
+ *
+ * <p>
+ * At this point, this rule is done, but further straightforward selection merging can rewrite the expression
+ * as:
+ * </p>
+ *
+ * <pre>{@code
+ * SELECT T.d
+ *   FROM T
+ *   WHERE a = 42 AND b = 'hello' AND c IS NULL
+ * }</pre>
+ */
+public class DecorrelateValuesRule extends ExplorationCascadesRule<SelectExpression> {
+
+    @Nonnull
+    private static final BindingMatcher<Quantifier.ForEach> forEach = QuantifierMatchers.forEachQuantifier();
+
+    @Nonnull
+    private static final BindingMatcher<SelectExpression> root = RelationalExpressionMatchers.selectExpression(some(forEach));
+
+    public DecorrelateValuesRule() {
+        super(root);
+    }
+
+    private static <T> boolean emptyIntersection(@Nonnull Set<? extends T> set1, @Nonnull Set<? super T> set2) {
+        return set1.stream().noneMatch(set2::contains);
+    }
+
+    private static boolean correlatedToNone(@Nonnull RelationalExpression expression, @Nonnull Set<CorrelationIdentifier> ids) {
+        return emptyIntersection(expression.getCorrelatedTo(), ids);
+    }
+
+    private static boolean correlatedToNone(@Nonnull Quantifier qun, @Nonnull Set<CorrelationIdentifier> ids) {
+        return emptyIntersection(qun.getCorrelatedTo(), ids);
+    }
+
+    @Override
+    public void onMatch(@Nonnull final ExplorationCascadesRuleCall call) {
+        final SelectExpression selectExpression = call.get(root);
+        final List<? extends Quantifier.ForEach> valueQunCandidates = call.getBindings().getAll(forEach);
+
+        final Set<CorrelationIdentifier> selectChildQunIds = selectExpression.getQuantifiers().stream().map(Quantifier::getAlias).collect(ImmutableSet.toImmutableSet());
+        ImmutableMap.Builder<CorrelationIdentifier, SelectExpression> valuesByIdBuilder = ImmutableMap.builderWithExpectedSize(valueQunCandidates.size());
+        ImmutableMap.Builder<CorrelationIdentifier, Quantifier> qunsByIdBuilder = ImmutableMap.builderWithExpectedSize(valueQunCandidates.size());
+        for (Quantifier.ForEach qun : valueQunCandidates) {
+            @Nullable SelectExpression childSelect = findSelectForQuantifier(qun, selectChildQunIds);
+            if (childSelect != null) {
+                valuesByIdBuilder.put(qun.getAlias(), childSelect);
+                qunsByIdBuilder.put(qun.getAlias(), qun);
+            }
+        }
+        final Map<CorrelationIdentifier, SelectExpression> valuesById = valuesByIdBuilder.build();
+        if (valuesById.isEmpty()) {
+            // No actual values boxes here. Exit now
+            return;
+        }
+        if (valuesById.size() == selectExpression.getQuantifiers().size()) {
```

Comment:
Isn't there a more general condition that you cannot find other quantifiers that are correlated to these ones so you cannot push at all?

Reply:
I'm not sure I understand the scenario here. If there are other quantifiers that are correlated to these, we can rewrite those other quantifiers so that the values box is now a child in that other quantifier's tree, I think